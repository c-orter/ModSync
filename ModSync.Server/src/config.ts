import path from "node:path";
import type { PreSptModLoader } from "@spt/loaders/PreSptModLoader";
import type { JsonUtil } from "@spt/utils/JsonUtil";
import type { VFS } from "@spt/utils/VFS";
import { glob, globNoEnd } from "./glob";
import { unixPath } from "./utility";
import type { ILogger } from "@spt/models/spt/utils/ILogger";

export type SyncPath = {
	path: string;
	enabled?: boolean;
	enforced?: boolean;
	silent?: boolean;
	restartRequired?: boolean;
};

type RawConfig = {
	syncPaths: (string | SyncPath)[];
	commonModExclusions: string[];
};

export class Config {
	constructor(
		public syncPaths: Required<SyncPath>[],
		public commonModExclusions: string[],
	) {}

	public isExcluded(filePath: string): boolean {
		return this.commonModExclusions.some((exclusion) =>
			glob(exclusion).test(unixPath(filePath)),
		);
	}

	public isParentExcluded(filePath: string): boolean {
		return this.commonModExclusions.some((exclusion) =>
			globNoEnd(exclusion).test(unixPath(filePath)),
		);
	}
}
export class ConfigUtil {
	constructor(
		private vfs: VFS,
		private jsonUtil: JsonUtil,
		private modImporter: PreSptModLoader,
		private logger: ILogger,
	) {}

	/**
	 * @throws {Error} If the config file does not exist
	 */
	private readConfigFile(): RawConfig {
		const modPath = this.modImporter.getModPath("Corter-ModSync");
		const configPath = path.join(modPath, "src", "config.jsonc");

		return this.jsonUtil.deserializeJsonC(
			this.vfs.readFile(configPath),
			"config.jsonc",
		) as RawConfig;
	}

	/**
	 * @throws {Error} If the config is invalid
	 */
	private validateConfig(config: RawConfig): void {
		if (!Array.isArray(config.syncPaths))
			throw new Error(
				"Corter-ModSync: config.jsonc 'syncPaths' is not an array. Please verify your config is correct and try again.",
			);

		if (!Array.isArray(config.commonModExclusions))
			throw new Error(
				"Corter-ModSync: config.jsonc 'commonModExclusions' is not an array. Please verify your config is correct and try again.",
			);

		for (const syncPath of config.syncPaths) {
			if (typeof syncPath !== "string" && !("path" in syncPath))
				throw new Error(
					"Corter-ModSync: config.jsonc 'syncPaths' is missing 'path'. Please verify your config is correct and try again.",
				);

			if (
				typeof syncPath === "string"
					? path.isAbsolute(syncPath)
					: path.isAbsolute(syncPath.path)
			)
				throw new Error(
					`Corter-ModSync: SyncPaths must be relative to SPT server root. Invalid path '${syncPath}'`,
				);

			if (
				path
					.relative(
						process.cwd(),
						path.resolve(
							process.cwd(),
							typeof syncPath === "string" ? syncPath : syncPath.path,
						),
					)
					.startsWith("..")
			)
				throw new Error(
					`Corter-ModSync: SyncPaths must within SPT server root. Invalid path '${syncPath}'`,
				);
		}
	}

	public load(): Config {
		const rawConfig = this.readConfigFile();
		this.validateConfig(rawConfig);

		return new Config(
			rawConfig.syncPaths.map((syncPath) => ({
				enabled: true,
				// Not yet implemented
				enforced: false,
				silent: false,
				restartRequired: true,
				...(typeof syncPath === "string" ? { path: syncPath } : syncPath),
			})),
			rawConfig.commonModExclusions,
		);
	}
}
