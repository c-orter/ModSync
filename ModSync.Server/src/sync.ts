import type { VFS } from "@spt/utils/VFS";
import path from "node:path";
import { crc32Init, crc32Update, crc32Final } from "./crc";
import type { Config, SyncPath } from "./config";
import { HttpError, winPath } from "./utility";
import type { ILogger } from "@spt/models/spt/utils/ILogger";
import { createReadStream } from "node:fs";

type ModFile = {
	crc: number;
	nosync: boolean;
};

export class SyncUtil {
	constructor(
		private vfs: VFS,
		private config: Config,
		private logger: ILogger,
	) {}

	private async getFilesInDir(dir: string): Promise<[string, boolean][]> {
		if (!this.vfs.exists(dir)) {
			this.logger.warning(
				`Corter-ModSync: Directory '${dir}' does not exist, will be ignored.`,
			);
			return [];
		}

		const stats = await this.vfs.statPromisify(dir);
		if (stats.isFile())
			return [
				[
					dir,
					this.config.isExcluded(dir) ||
						this.vfs.exists(path.join(dir, ".nosync")) ||
						this.vfs.exists(path.join(dir, ".nosync.txt")),
				],
			];

		const nosyncDir =
			this.config.isExcluded(dir) ||
			this.vfs.exists(path.join(dir, ".nosync")) ||
			this.vfs.exists(path.join(dir, ".nosync.txt"));

		return (
			await Promise.all(
				this.vfs
					.getFiles(dir)
					.filter(
						(file) =>
							!file.endsWith(".nosync") && !file.endsWith(".nosync.txt"),
					)
					.map(
						async (file): Promise<[string, boolean]> => [
							path.join(dir, file),
							nosyncDir ||
								this.config.isExcluded(path.join(dir, file)) ||
								this.vfs.exists(`${path.join(dir, file)}.nosync`) ||
								this.vfs.exists(`${path.join(dir, file)}.nosync.txt`),
						],
					),
			)
		).concat(
			(
				await Promise.all(
					this.vfs
						.getDirs(dir)
						.map((subDir) => this.getFilesInDir(path.join(dir, subDir))),
				)
			)
				.flat()
				.map(([child, nosync]): [string, boolean] => [
					child,
					nosyncDir || nosync,
				]),
		);
	}

	private async buildModFile(
		file: string,
		// biome-ignore lint/correctness/noEmptyPattern: <explanation>
		{}: Required<SyncPath>,
		nosync: boolean,
	): Promise<ModFile> {
		try {
			return {
				nosync,
				crc: nosync ? 0 : await new Promise<number>((resolve, reject) => {
					let crc = crc32Init();

					createReadStream(file)
						.on("error", reject)
						.on("data", (data: Buffer) => {
							crc = crc32Update(crc, data);
						})
						.on("end", () => {
							resolve(crc32Final(crc));
						});
				}),
			};
		} catch (e) {
			throw new HttpError(500, `Corter-ModSync: Error reading '${file}'\n${e}`);
		}
	}

	public async hashModFiles(
		syncPaths: Config["syncPaths"],
	): Promise<Record<string, Record<string, ModFile>>> {
		return Object.fromEntries(
			await Promise.all(
				syncPaths.map(async (syncPath) => [
					winPath(syncPath.path),
					Object.fromEntries(
						await Promise.all(
							(await this.getFilesInDir(syncPath.path)).map(
								async ([file, nosync]) =>
									[
										winPath(file),
										await this.buildModFile(file, syncPath, nosync),
									] as const,
							),
						),
					),
				]),
			),
		);
	}

	/**
	 * @throws {Error} If file path is invalid
	 */
	public sanitizeDownloadPath(
		file: string,
		syncPaths: Config["syncPaths"],
	): string {
		const normalized = path.join(
			path.normalize(file).replace(/^(\.\.(\/|\\|$))+/, ""),
		);

		if (
			!syncPaths.some(
				({ path: p }) =>
					!path
						.relative(path.join(process.cwd(), p), normalized)
						.startsWith(".."),
			)
		)
			throw new HttpError(
				400,
				`Corter-ModSync: Requested file '${file}' is not in an enabled sync path!`,
			);

		return normalized;
	}
}
