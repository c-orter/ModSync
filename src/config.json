{
    "clientDirs": ["BepInEx/plugins", "BepInEx/config"],
    "serverDirs": ["user/mods"]
}