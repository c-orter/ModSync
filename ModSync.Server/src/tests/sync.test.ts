import { expect, describe, it, vi, beforeEach } from "vitest";
import { fs, vol } from "memfs";

import { SyncUtil } from "../sync";
import { Config } from "../config";
import { VFS } from "./utils/vfs";
import type { VFS as IVFS } from "@spt/utils/VFS";
import type { ILogger } from "@spt/models/spt/utils/ILogger";
import { mock } from "vitest-mock-extended";

vi.mock("node:fs", () => ({ ...fs, default: fs }));

const directoryStructure = {
	plugins: {
		"file1.dll": "Test content",
		"file2.dll": "Test content 2",
		"file2.dll.nosync": "",
		"file3.dll": "Test content 3",
		"file3.dll.nosync.txt": "",
		ModName: {
			"mod_name.dll": "Test content 4",
			".nosync": "",
		},
		OtherMod: {
			"other_mod.dll": "Test content 5",
			subdir: {
				"image.png": "Test Image",
				".nosync": "",
			},
		},
	},
	user: {
		mods: {
			testMod: {
				"test.ts": "Test content 6",
				"test.js": "Test content 6",
				"test.js.map": "Test content 7",
			},
		},
	},
};

const config = new Config(
	[
		{
			path: "plugins",
			enabled: true,
			enforced: false,
			restartRequired: true,
			silent: false,
		},
	],
	[],
);

const vfs = new VFS();
const logger = mock<ILogger>();
const syncUtil = new SyncUtil(vfs as IVFS, config, logger);

beforeEach(() => {
	vol.reset();
	vol.fromNestedJSON(directoryStructure);
});

describe("hashModFiles", () => {
	it("should hash mod files", () => {
		const hashes = syncUtil.hashModFiles(config.syncPaths);

		expect(
			Object.keys(hashes).some(
				(file) => file.endsWith(".nosync") || file.endsWith(".nosync.txt"),
			),
		).toBe(false);

		expect("plugins\\OtherMod\\other_mod.dll" in hashes["plugins"]).toBe(true);

		expect(hashes).toMatchSnapshot();
	});

	it("should correctly hash multiple folders", () => {
		const config = new Config(
			[
				{
					path: "plugins",
					enabled: true,
					enforced: false,
					restartRequired: true,
					silent: false,
				},
				{
					path: "user/mods",
					enabled: true,
					enforced: false,
					restartRequired: false,
					silent: false,
				},
			],
			["user/mods/**/*.js", "user/mods/**/*.js.map"],
		);

		const vfs = new VFS();
		const syncUtil = new SyncUtil(vfs as IVFS, config, logger);

		const hashes = syncUtil.hashModFiles(config.syncPaths);

		expect(hashes).toMatchSnapshot();
	});

	it("should correctly ignore folders that do not exist", () => {
		const config = new Config(
			[
				{
					path: "plugins",
					enabled: true,
					enforced: false,
					restartRequired: true,
					silent: false,
				},
				{
					path: "user/bananas",
					enabled: true,
					enforced: false,
					restartRequired: false,
					silent: false,
				},
			],
			["user/mods/**/*.js", "user/mods/**/*.js.map"],
		);

		const vfs = new VFS();
		const syncUtil = new SyncUtil(vfs as IVFS, config, logger);

		const hashes = syncUtil.hashModFiles(config.syncPaths);

		expect(Object.keys(hashes["plugins"])).toContain("plugins\\file1.dll");
		expect(Object.keys(hashes["plugins"])).toContain(
			"plugins\\OtherMod\\other_mod.dll",
		);
		expect(logger.warning).toHaveBeenCalledWith(
			"Corter-ModSync: Directory 'user/bananas' does not exist, will be ignored.",
		);
	});

	it("should correctly hash folders that didn't exist initially but are created", () => {
		const config = new Config(
			[
				{
					path: "plugins",
					enabled: true,
					enforced: false,
					restartRequired: true,
					silent: false,
				},
				{
					path: "user/bananas",
					enabled: true,
					enforced: false,
					restartRequired: false,
					silent: false,
				},
			],
			["user/mods/**/*.js", "user/mods/**/*.js.map"],
		);

		const vfs = new VFS();
		const syncUtil = new SyncUtil(vfs as IVFS, config, logger);

		const hashes = syncUtil.hashModFiles(config.syncPaths);

		expect(Object.keys(hashes["plugins"])).toContain("plugins\\file1.dll");
		expect(Object.keys(hashes["plugins"])).toContain(
			"plugins\\OtherMod\\other_mod.dll",
		);

		expect(logger.warning).toHaveBeenCalledWith(
			"Corter-ModSync: Directory 'user/bananas' does not exist, will be ignored.",
		);

		fs.mkdirSync("user/bananas", { recursive: true });
		fs.writeFileSync("user/bananas/test.txt", "test");

		const newHashes = syncUtil.hashModFiles(config.syncPaths);

		expect(newHashes).toMatchSnapshot();
	});
});

describe("sanitizeDownloadPath", () => {
	it("should sanitize correct download paths", () => {
		expect(
			syncUtil.sanitizeDownloadPath("plugins\\file1.dll", config.syncPaths),
		).toBe("plugins\\file1.dll");
	});

	it("should throw for download paths outside SPT root", () => {
		expect(() => {
			syncUtil.sanitizeDownloadPath("plugins\\..\\file1.dll", config.syncPaths);
		}).toThrow();
	});

	it("should throw for files not in syncPath", () => {
		expect(() => {
			syncUtil.sanitizeDownloadPath("otherDir\\file1.dll", config.syncPaths);
		}).toThrow();
	});
});
