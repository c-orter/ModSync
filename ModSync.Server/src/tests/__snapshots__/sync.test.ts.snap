// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hashModFiles > should correctly hash folders that didn't exist initially but are created 1`] = `
{
  "plugins": {
    "plugins\\ModName\\mod_name.dll": {
      "crc": 0,
      "nosync": true,
    },
    "plugins\\OtherMod\\other_mod.dll": {
      "crc": 2471037616,
      "nosync": false,
    },
    "plugins\\OtherMod\\subdir\\image.png": {
      "crc": 0,
      "nosync": true,
    },
    "plugins\\file1.dll": {
      "crc": 1338358949,
      "nosync": false,
    },
    "plugins\\file2.dll": {
      "crc": 0,
      "nosync": true,
    },
    "plugins\\file3.dll": {
      "crc": 0,
      "nosync": true,
    },
  },
  "user\\bananas": {
    "user\\bananas\\test.txt": {
      "crc": 3632233996,
      "nosync": false,
    },
  },
}
`;

exports[`hashModFiles > should correctly hash multiple folders 1`] = `
{
  "plugins": {
    "plugins\\ModName\\mod_name.dll": {
      "crc": 0,
      "nosync": true,
    },
    "plugins\\OtherMod\\other_mod.dll": {
      "crc": 2471037616,
      "nosync": false,
    },
    "plugins\\OtherMod\\subdir\\image.png": {
      "crc": 0,
      "nosync": true,
    },
    "plugins\\file1.dll": {
      "crc": 1338358949,
      "nosync": false,
    },
    "plugins\\file2.dll": {
      "crc": 0,
      "nosync": true,
    },
    "plugins\\file3.dll": {
      "crc": 0,
      "nosync": true,
    },
  },
  "user\\mods": {
    "user\\mods\\testMod\\test.js": {
      "crc": 0,
      "nosync": true,
    },
    "user\\mods\\testMod\\test.js.map": {
      "crc": 0,
      "nosync": true,
    },
    "user\\mods\\testMod\\test.ts": {
      "crc": 171989770,
      "nosync": false,
    },
  },
}
`;

exports[`hashModFiles > should hash mod files 1`] = `
{
  "plugins": {
    "plugins\\ModName\\mod_name.dll": {
      "crc": 0,
      "nosync": true,
    },
    "plugins\\OtherMod\\other_mod.dll": {
      "crc": 2471037616,
      "nosync": false,
    },
    "plugins\\OtherMod\\subdir\\image.png": {
      "crc": 0,
      "nosync": true,
    },
    "plugins\\file1.dll": {
      "crc": 1338358949,
      "nosync": false,
    },
    "plugins\\file2.dll": {
      "crc": 0,
      "nosync": true,
    },
    "plugins\\file3.dll": {
      "crc": 0,
      "nosync": true,
    },
  },
}
`;
